File: Remove K Digits.cpp
// leetcode 402 problem
// 在字符串中移除k个字符，使得余下的字符组成的数最小

// Straight-forward O(k*n) Solution.
class Solution {
public:
    string removeKdigits(string num, int k) {
        // 每次移除从左往右第一个比右边大的数
        while (k > 0) {
            int n = num.size();
            int i = 0;
            while (i+1 < n && num[i] <= num[i+1]) i++;
            num.erase(i, 1);
            k--;
        }
        // 移除开头的0
        int s = 0;
        while (s < (int)num.size()-1 && num[s] == '0') s++;
        num.erase(0, s);
        return num == "" ? "0" : num;
    }
};

// O(n) Solution(using stack).
class Solution {
public:
    string removeKdigits(string num, int k) {
        string res;
        int keep = num.size() - k;
        for (int i = 0; i < num.size(); i++) {
            while (res.size() > 0 && res.back() > num[i] && k > 0) {
                res.pop_back();
                k--;
            }
            res.push_back(num[i]);
        }
        // prevent only 1 char case
        res.erase(keep, string::npos);
        // trim leading zeros
        int s = 0;
        while (s < (int)res.size()-1 && res[s] == '0') s++;
        res.erase(0, s);
        return res == "" ? "0" : res;
    }
};

File: Remove K Digits.java
// leetcode 402 problem
// 移除字符串中的k个字符，使余下的数字最小

class Solution {
    public String removeKdigits(String num, int k) {
        if (num == null || k == 0) return num;
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < num.length(); i++) {
            while (k > 0 && !stack.isEmpty() && stack.peek() > num.charAt(i)) {
                stack.pop();
                k--;
            }
            stack.push(num.charAt(i));
        }

        // Case 1111.
        while (k > 0) {stack.pop(); k--;}

        // Build String.
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        sb.reverse();

        // Get rid of leading 0.
        while (sb.length() > 1 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }
        return sb.toString();
    }
}